### PR TITLE
MAINT/ENH: add SiLU activation function

### PR DESCRIPTION
**Pull request by nghi-truyen**
NaN gradients are produced when using a neural ODE solver on large datasets. The issue appears to stem from the differentiability of the process-parameterization NN (which must be twice differentiable mathematically) and the structure of the Jacobian matrix in the Fortran code.

This PR resolves the issue, tested on ArcMed and Aude catchments, with the following changes:
- The SiLU activation function (which is twice differentiable) replaces ReLU (non-differentiable at 0) in the process-parameterization NN. Since the initial conditions make the output layers very close to zero, using ReLU leads to issues with differentiability at this point;
- SiLU is also added to the list of activation functions available in the Net subpackage;
- Two vectors are used instead of a matrix for computing the Jacobian in the backward routine of the process-parameterization NN.

Baseline:
- The tests for simulated discharges use higher tolerance due to machine precision problems.
- Thus, we check only high flows (that exceed the 0.95-quantile).

### CODE REVIEW

File: smash/factory/net/_layers.py [hunk 1]
```diff
@@ -393,6 +393,15 @@ def gradient(self, x):
...
 
 
+class SiLU:
+    def __call__(self, x):
+        return x / (1 + np.exp(-x))
+
+    def gradient(self, x):
+        s = 1 / (1 + np.exp(-x))
+        return s + x * s * (1 - s)
+
```
Comment (pag13) [inline smash/factory/net/_layers.py:403]: ok for Python function

File: smash/fcore/operator/md_neural_network.f90 [hunk 2]
```diff
@@ -75,53 +76,53 @@ subroutine forward_and_backward_mlp(weight_1, bias_1, weight_2, bias_2, weight_3
...
-        layer_2_gradient = 0._sp
+        output_jacobian_1 = 0._sp
+        output_jacobian_2 = 0._sp
 
         call dot_product_2d_1d(weight_1, input_layer, inter_layer_1)
         inter_layer_1 = inter_layer_1 + bias_1
-        inter_layer_1_tf = max(0.01_sp*inter_layer_1, inter_layer_1) ! Leaky ReLU
+        inter_layer_1_tf = inter_layer_1*(1._sp/(1._sp + exp(-inter_layer_1))) ! SiLU
+        inter_layer_1_grad = inter_layer_1_tf + &
+        & (1._sp - inter_layer_1_tf)/(1._sp + exp(-inter_layer_1))  ! Derivative of SiLU
```
Comment (pag13) [inline smash/fcore/operator/md_neural_network.f90:93]: grad = s + x * s * (1 - s)  avec s = 1 / (1 + exp(-x)) ?
manque * x pour le second terme ?

File: smash/fcore/operator/md_neural_network.f90 [hunk 3]
```diff
@@ -75,53 +76,53 @@ subroutine forward_and_backward_mlp(weight_1, bias_1, weight_2, bias_2, weight_3
...
 
         if (size(bias_3) .gt. 0) then  ! Case with 3 layers
             call dot_product_2d_1d(weight_2, inter_layer_1_tf, inter_layer_2)
             inter_layer_2 = inter_layer_2 + bias_2
-            inter_layer_2_tf = max(0.01_sp*inter_layer_2, inter_layer_2) ! Leaky ReLU
+            inter_layer_2_tf = inter_layer_2*(1._sp/(1._sp + exp(-inter_layer_2))) ! SiLU
+            inter_layer_2_grad = inter_layer_2_tf + &
+            & (1._sp - inter_layer_2_tf)/(1._sp + exp(-inter_layer_2))  ! Derivative of SiLU
```
Comment (pag13) [inline smash/fcore/operator/md_neural_network.f90:100]: idem

File: smash/tests/core/simulation/test_bayesian_optimize.py [hunk 4]
```diff
@@ -76,7 +76,7 @@ def generic_custom_bayesian_optimize(model: smash.Model, **kwargs) -> dict:
         instance = smash.bayesian_optimize(model, **inner_kwargs)
 
         qsim = instance.response.q[:].flatten()
-        qsim = qsim[::10]  # extract values at every 10th position
+        qsim = qsim[qsim > np.quantile(qsim, 0.95)]  # extract values depassing 0.95-quantile
```
Comment (pag13) [inline smash/tests/core/simulation/test_bayesian_optimize.py:79]: modif de test_bayesian_optimize nécessaire vs solver ode-hybride ?

**Review by pag13 — COMMENTED** [review]
Ok comme cela, en confirmant l'expression du gradient silu implementé en fortran

File: smash/factory/net/net.py [hunk 5]
```diff
@@ -203,6 +203,7 @@ def add_dense(
             - ``'leakyrelu'`` : Leaky Rectified Linear Unit
             - ``'tanh'`` : Hyperbolic Tangent
             - ``'softplus'`` : Softplus
+            - ``'silu'`` : Sigmoid Linear Unit
```
[the remark this answers is not in the document]
Comment (nghi-truyen) [reply smash/factory/net/net.py:207]: Je sais pas trop, les deux marchent mais je trouve que "Sigmoid Linear Unit" est souvent utilisé:
https://pytorch.org/docs/stable/generated/torch.nn.SiLU.html
